### PR TITLE
[api-gateway]: Support http(s) as AppProtocol in Kubernetes svc

### LINKED ISSUE #6560: Service AppProtocol should allow IANA standard service names like "http", "https"
**What steps did you take and what happened:**

Create a service with appProtocol:

```
apiVersion: v1
kind: Service
metadata:
  name: hello-service
spec:
  ports:
  - appProtocol: https
    name: https
    port: 443
    protocol: TCP
    targetPort: https
```

Create a httproute with pointing to to service:

```
apiVersion: gateway.networking.k8s.io/v1
kind: HTTPRoute
metadata:
  name: hello-route
spec:
  parentRefs:
    - group: gateway.networking.k8s.io
      kind: Gateway
      namespace: projectcontour
      name: contour
      sectionName: https
  hostnames:
    - example.com
  rules:
    - backendRefs:
      - kind: Service
        name: hello-service
        port: 80

```

**What did you expect to happen:**

The route is installed successfully, but got an error:

```
    - lastTransitionTime: "2024-07-17T09:36:00Z"
      message: 'AppProtocol: "http" is unsupported'
      observedGeneration: 3
      reason: UnsupportedProtocol
      status: "False"
      type: ResolvedRefs
```

According to the spec https://kubernetes.io/docs/concepts/services-networking/service/#application-protocol

appProtocol can be IANA service names listed here: https://www.iana.org/assignments/service-names-port-numbers/service-names-port-numbers.xhtml

These should be all valid:
* http
* www
* www-http
* https

**Anything else you would like to add:**

**Environment:**

- Contour version: v1.29.1
- Kubernetes version: (use `kubectl version`): v1.30.1
- Kubernetes installer & version: 
- Cloud provider or hardware configuration:
- OS (e.g. from `/etc/os-release`):

### PR DESCRIPTION
**Pull request by Krast76**
When contour is used as api-gateway in kubernetes clusters, appProtocol set to http or https doesn't work.
This PR, if applied, allow the following protocol (as defined by [IANA](https://www.iana.org/assignments/service-names-port-numbers/service-names-port-numbers.xhtml?search=https)) : 

- http and www-http
- https

That doesn't allow all the IANA services, but that'll support a common case

Fix #6560

### CODE REVIEW

File: internal/dag/accessors.go [hunk 1]
```diff
@@ -108,15 +108,21 @@ func validateExternalName(svc *core_v1.Service, enableExternalNameSvc bool) erro
 
 // the ServicePort's AppProtocol must be one of the these.
 const (
-	protoK8sH2C = "kubernetes.io/h2c"
-	protoK8sWS  = "kubernetes.io/ws"
+	protoK8sH2C  = "kubernetes.io/h2c"
+	protoK8sWS   = "kubernetes.io/ws"
+	protoHttp    = "www-http"
```
Comment (tsaarni) [inline internal/dag/accessors.go:113]: Should `www-http` be removed?

It seems to be historical name, as well as `www` which is could also be found in IANA registry, https://datatracker.ietf.org/doc/html/rfc6335#section-5
> By historical accident, the service name "http" has two synonyms
      "www" and "www-http".  When used in SRV records [[RFC2782](https://datatracker.ietf.org/doc/html/rfc2782)] and
      similar service discovery mechanisms, only the service name "http"
      should be used, not these additional names.

Comment (Krast76) [reply internal/dag/accessors.go:113]: I didn't see that, thanks !
I'm removing it

**Comment by tsaarni** [comment]
Thanks @Krast76! Could you check also the linter nags here https://github.com/projectcontour/contour/pull/6616/files and add a [changelog](https://github.com/projectcontour/contour/tree/main/changelogs) entry in the PR, the filename should be `changelogs/unreleased/6616-Krast76-small.md`.

**Comment by Krast76** [comment]
> Thanks @Krast76! Could you check also the linter nags here https://github.com/projectcontour/contour/pull/6616/files and add a [changelog](https://github.com/projectcontour/contour/tree/main/changelogs) entry in the PR, the filename should be `changelogs/unreleased/6616-Krast76-small.md`.

Done :)

**Comment by Krast76** [comment]
@tsaarni There is something else I should do to make the PR merged ?

The linter still failed, but I don't really understand why. 🤔

**Comment by tsaarni** [comment]
> The linter still failed, but I don't really understand why. 🤔

Almost impossible to see in the error, but it was the indentation that it complains :-)  Run `make format` to get it fixed!

**Comment by tsaarni** [comment]
> @tsaarni There is something else I should do to make the PR merged ?

It could be good to add a unit test, something like this

```diff
diff --git a/internal/dag/accessors_test.go b/internal/dag/accessors_test.go
index bc3e2b86..8c3bc553 100644
--- a/internal/dag/accessors_test.go
+++ b/internal/dag/accessors_test.go
@@ -127,6 +127,18 @@ func TestBuilderLookupService(t *testing.T) {
                                        AppProtocol: ptr.To("kubernetes.io/wss"),
                                        Port:        8444,
                                },
+                               {
+                                       Name:        "iana-https",
+                                       Protocol:    "TCP",
+                                       AppProtocol: ptr.To("https"),
+                                       Port:        8445,
+                               },
+                               {
+                                       Name:        "iana-http",
+                                       Protocol:    "TCP",
+                                       AppProtocol: ptr.To("http"),
+                                       Port:        8446,
+                               },
                        },
                },
        }
@@ -212,12 +224,21 @@ func TestBuilderLookupService(t *testing.T) {
                        port:           8443,
                        want:           appProtcolService(appProtoService, "h2c"),
                },
-
                "lookup service by port number with unsupported k8s app protocol: wss": {
                        NamespacedName: types.NamespacedName{Name: appProtoService.Name, Namespace: appProtoService.Namespace},
                        port:           8444,
                        want:           appProtcolService(appProtoService, "", 1),
                },
+               "lookup service by port number with supported IANA app protocol: https": {
+                       NamespacedName: types.NamespacedName{Name: appProtoService.Name, Namespace: appProtoService.Namespace},
+                       port:           8445,
+                       want:           appProtcolService(appProtoService, "tls", 2),
+               },
+               "lookup service by port number with supported IANA app protocol: http": {
+                       NamespacedName: types.NamespacedName{Name: appProtoService.Name, Namespace: appProtoService.Namespace},
+                       port:           8446,
+                       want:           appProtcolService(appProtoService, "", 3),
+               },
        }

        for name, tc := range tests {
```

**Comment by Krast76** [comment]
> > The linter still failed, but I don't really understand why. 🤔
> 
> Almost impossible to see in the error, but it was the alignment that it complains :-) Run `make format` to get it fixed!

I should have guessed that 😅 

The linter must be ok (for real) this time, and I have added the unit tests you have suggested, thanks ! :)

**Review by tsaarni — APPROVED** [review]
Thank You @Krast76!

**Pull request merged**